nada_dsl/nada_types/scalar_types.py

```diff
@@ -0,0 +1,458 @@
+        return BaseType.UNSIGNED_INTEGER
+
+    def to_mode(self) -> Mode:
+        return Mode.PUBLIC
+
+    def __eq__(self, other):
+        return ScalarType.__eq__(self, other)
+
+    def public_equals(self, other: Union["PublicUnsignedInteger", "SecretUnsignedInteger"]) -> "PublicBoolean":
+        return public_equals_operation(self, other)
+
+
+@dataclass
+class PublicBoolean(BooleanType):
+
+    def __init__(self, inner: NadaType):
+        super().__init__(inner)
+
+    def to_base_type(self) -> BaseType:
+        return BaseType.BOOLEAN
+
+    def to_mode(self) -> Mode:
+        return Mode.PUBLIC
+
+    def __eq__(self, other):
+        return ScalarType.__eq__(self, other)
+
+    def __invert__(self: "PublicBoolean") -> "PublicBoolean":
+        operation = Not(this=self, source_ref=SourceRef.back_frame())
+        return PublicBoolean(inner=operation)
+
+
+@dataclass
+class SecretInteger(NumericType):
+
+    def __init__(self, inner: NadaType):
+        super().__init__(inner)
+
+    def to_base_type(self) -> BaseType:
+        return BaseType.INTEGER
+
+    def to_mode(self) -> Mode:
+        return Mode.SECRET
+
+    def __eq__(self, other):
+        return ScalarType.__eq__(self, other)
+
+    def public_equals(self, other: Union["PublicInteger", "SecretInteger"]) -> "PublicBoolean":
+        return public_equals_operation(self, other)
+
+    def trunc_pr(self, other: Union["PublicUnsignedInteger", "UnsignedInteger"]) -> "SecretInteger":
+        if isinstance(other, UnsignedInteger):
+            operation = TruncPr(left=self, right=other, source_ref=SourceRef.back_frame())
+            return SecretInteger(inner=operation)
+        elif isinstance(other, PublicUnsignedInteger):
+            operation = TruncPr(left=self, right=other, source_ref=SourceRef.back_frame())
+            return SecretInteger(inner=operation)
+        else:
+            raise TypeError(f"Invalid operation: {self}.trunc_pr({other})")
+
+    @classmethod
+    def random(cls) -> "SecretInteger":
+        return SecretInteger(inner=Random(source_ref=SourceRef.back_frame()))
+
+    def to_public(self: "SecretInteger") -> "PublicInteger":
+        operation = Reveal(this=self, source_ref=SourceRef.back_frame())
+        return PublicInteger(inner=operation)
+
+
+@dataclass
+class SecretUnsignedInteger(NumericType):
+    def __init__(self, inner: NadaType):
+        super().__init__(inner)
+
+    def to_base_type(self) -> BaseType:
+        return BaseType.UNSIGNED_INTEGER
+
+    def to_mode(self) -> Mode:
+        return Mode.SECRET
+
+    def __eq__(self, other):
+        return ScalarType.__eq__(self, other)
+
+    def public_equals(self, other: Union["PublicUnsignedInteger", "SecretUnsignedInteger"]) -> "PublicBoolean":
+        return public_equals_operation(self, other)
+
+    def trunc_pr(self, other: Union["PublicUnsignedInteger", "UnsignedInteger"]) -> "SecretUnsignedInteger":
+        if isinstance(other, UnsignedInteger):
```

Comment:
`trunc_pr` is only implemented for `SecretInteger` and `SecretUnsignedInteger`. If I move it to `NumericType`, it will be available in the `Public` and `Constant` modes. I can check the types in the function, but the IDEs would allow using it and that would be a bit confusing.